lib/_stream_writable.js

```diff
@@ -362,27 +388,34 @@ function onwriteDrain(stream, state) {
 function clearBuffer(stream, state) {
   state.bufferProcessing = true;
   var entry = state.bufferedRequest;
+  var bufferedRequests = state.bufferedRequests;
+
+  state.bufferedRequests = 0;
 
   if (stream._writev && entry && entry.next) {
     // Fast case, write everything using _writev()
-    var buffer = [];
-    var cbs = [];
+    var buffer = new Array(bufferedRequests);
+    var cbs = new Array(bufferedRequests);
+    var count = 0;
+
     while (entry) {
-      cbs.push(entry.callback);
-      buffer.push(entry);
+      cbs[count] = entry.callback;
+      buffer[count] = entry;
+      count++;
       entry = entry.next;
     }
 
     // count the one we are adding, as well.
-    // TODO(isaacs) clean this up
     state.pendingcb++;
     state.lastBufferedRequest = null;
-    doWrite(stream, state, true, state.length, buffer, '', function(err) {
-      for (var i = 0; i < cbs.length; i++) {
-        state.pendingcb--;
-        cbs[i](err);
-      }
-    });
+
+    if (state.corkedCbs) {
+      state.corkedCbs.next = new CorkCbs(cbs)
```

Comment:
@chrisdickinson good spot! I planned to write more about this! My understanding is that it is only 2 `corkedCbs` long. Maybe I am wrong, see these lines, the callback can be deferred on the nextTick: https://github.com/mcollina/node/blob/perf-clearBuffer/lib/_stream_writable.js#L360-L364.

There might also be a different way to implement that, maybe having an array with two objects, and a even/odd counter.